src/test/java/com/puppycrawl/tools/checkstyle/meta/MetadataGenerator.java

```diff
@@ -0,0 +1,99 @@
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.meta;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.puppycrawl.tools.checkstyle.Checker;
+import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
+import com.puppycrawl.tools.checkstyle.TreeWalker;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+
+public final class MetadataGenerator {
+
+    private MetadataGenerator() {
+    }
+
+    /**
+     * Generate metadata from the module source files available in the input argument path.
+     *
+     * @param args arguments
+     * @throws IOException ioException
+     * @throws CheckstyleException checkstyleException
+     */
+    public static void generate(String... args) throws IOException, CheckstyleException {
+        final Checker checker = new Checker();
+        checker.setModuleClassLoader(Checker.class.getClassLoader());
+        final DefaultConfiguration scraperCheckConfig =
+                        new DefaultConfiguration(JavadocMetadataScraper.class.getName());
+        final DefaultConfiguration defaultConfiguration = new DefaultConfiguration("configuration");
+        final DefaultConfiguration treeWalkerConfig =
+                new DefaultConfiguration(TreeWalker.class.getName());
+        defaultConfiguration.addAttribute("charset", StandardCharsets.UTF_8.name());
+        defaultConfiguration.addChild(treeWalkerConfig);
+        treeWalkerConfig.addChild(scraperCheckConfig);
+        checker.configure(defaultConfiguration);
+        dumpMetadata(checker, args[0]);
+    }
+
+    /**
+     * Process files using the checker passed and write to corresponding XML files.
+     *
+     * @param checker checker
+     * @param path rootPath
+     * @throws CheckstyleException checkstyleException
+     * @throws IOException ioException
+     */
+    private static void dumpMetadata(Checker checker, String path) throws CheckstyleException,
+            IOException {
+        final List<File> validFiles = new ArrayList<>();
+        if (path.endsWith(".java")) {
+            validFiles.add(new File(path));
+        }
+        else {
+            final List<String> moduleFolders = Arrays.asList("checks", "filters", "filefilters");
+            for (String folder : moduleFolders) {
+                try (Stream<Path> files = Files.walk(Paths.get(path
+                        + "/" + folder))) {
+                    validFiles.addAll(files.map(Path::toString)
```

Comment:
`files.map(Path::toString)`
there is no need to convert Path into string and then into file, Path has methods `boolean endsWith(String other)` and  `File toFile()` which will make it simpler

Please correct other occurrences if any

Reply:
Done